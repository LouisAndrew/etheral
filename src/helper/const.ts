import { Product } from './schema/product';
import { Order } from './schema/order';
import { Blog } from './schema/blog';
import { Currencies } from 'state/reducers/currency-reducer';

export const mockProduct1: Product = {
    pid: '1234',
    name: 'Mock product',
    slug: 'ctg1/product1',
    category: 'ctg1',
    productDetails: '<p>Good product</p>',
    description: 'Just a mock product',
    prices: {
        idrPrice: 23000,
        ausPrice: 21,
        discountPercentage: 10,
    },
    gems: {
        withGems: true,
        gemTypes: 'A, B',
        gemSizes: 'S, M, L',
    },
    availableSizes: 'S, M, L',
    collection: 'fall',
    productImages: [],
    urls: [],
    weight: 20,
    amount: 12,
    relatedProducts: [],
};

export const mockProducts: Product[] = [mockProduct1];

export const mockOrder: Order = {
    oid: 'asdadwda',
    buyerName: 'Jane Doe',
    buyerEmail: 'jane@doe',
    buyerPhone: 123412412,
    buyerUId: 'admin',
    buyerAddr: 'Storkowerstrasse',
    buyerPostal: 1234143,
    total: 32000,
    currency: Currencies.IDR,
    date: new Date(),
    via: 'va',
    channel: 'bni',
    products: [
        {
            pid: mockProduct1.pid,
            amount: 12,
            discountPercentage: 10,
            note: [
                {
                    details: {
                        size: 'S',
                    },
                    amount: 12,
                },
            ],
        },
    ],
    paid: false,
    delivered: false,
    transactionData: {
        sessionId: '1234',
        paymentNo: '12',
        paymentName: 'Jane Doe - BNI',
        expired: '1231241412',
        fee: 3000,
    },
    shippingMethod: 'OKE',
};

export const mockOrders: Order[] = [mockOrder];

export const mockOrderShipped: Order = {
    ...mockOrder,
    delivered: true,
    shippingData: {
        shippingDate: new Date(),
        trackingNum: '12342',
        shippedBy: 'admin1',
    },
};

export const mockBlog: Blog = {
    content:
        '<h1 role="heading">Hello World!</h1><br /><img src="" alt="empty image" />',
    summary: 'Mock summary',
    title: 'Mock title',
    date: new Date(Date.parse('2020-09-11T00:00:00+02:00')),
    slug: 'sample-blog-post',
};

export const nameToSlug = (name: string) =>
    name
        .toLowerCase()
        .split(' ')
        .join('-');
