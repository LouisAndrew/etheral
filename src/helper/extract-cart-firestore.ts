import { intersectionBy } from 'lodash';

import { SetCartArgs } from 'state/actions/cart';
import { InCart } from './schema/firebase-user';
import { Product } from './schema/product';

export default ({
    firestoreCartData,
    allProducts,
}: {
    firestoreCartData: InCart;
    allProducts: Product[];
}): SetCartArgs => {
    const inCartProduct = intersectionBy(allProducts, firestoreCartData, 'pid');
    return inCartProduct.map(product => ({
        product,
        amount: intersectionBy(firestoreCartData, [product], 'pid')[0].amount, // get the intersection between the product -> pid is unique, so the function is guaranteed to return one item.
    }));
};
