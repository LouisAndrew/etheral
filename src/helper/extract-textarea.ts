import { snakeCase } from 'lodash';

/**
 * function to extract options from a textarea
 * @param textarea textarea input (e.g sizes: S, M, L)
 */
export const extractTextArea = (textarea: string) =>
    textarea.split(',').map(text => snakeCase(text).toUpperCase());
