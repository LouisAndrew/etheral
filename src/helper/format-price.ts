export const formatPrice = (price: number | string) => {
    const value =
        typeof price === 'string'
            ? parseFloat(price.replace('IDR ', ''))
            : price;
    return new Intl.NumberFormat('id').format(value);
};
