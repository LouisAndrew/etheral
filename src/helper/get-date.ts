const pad2 = (n: number): string => (n < 10 ? `0${n}` : n.toString());

const getDate = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = pad2(date.getUTCMonth() + 1); // getMonth() is zero-indexed, so we'll increment to get the correct month number
    const day = pad2(date.getUTCDate());
    const hours = pad2(date.getUTCHours() + 7);
    const minutes = pad2(date.getUTCMinutes());

    return year + '-' + month + '-' + day + ' ' + hours + ':' + minutes;
};

const getDateIpaymu = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = pad2(date.getUTCMonth() + 1); // getMonth() is zero-indexed, so we'll increment to get the correct month number
    const day = pad2(date.getUTCDate());
    const hours = pad2(date.getUTCHours() + 7); // UTC offset for WIB => +7
    const minutes = pad2(date.getUTCMinutes());
    const seconds = pad2(date.getUTCSeconds());

    return `${year}${month}${day}${hours}${minutes}${seconds}`;
};

const monthNames = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
];

const getDateReadable = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = date.getUTCMonth();
    const day = date.getUTCDate();

    return `${day} ${monthNames[month]} ${year}`;
};

export { getDate, getDateIpaymu, getDateReadable };
