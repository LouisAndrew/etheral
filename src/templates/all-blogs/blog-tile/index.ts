
import { BlogTile } from './blog-tile';

export default BlogTile;