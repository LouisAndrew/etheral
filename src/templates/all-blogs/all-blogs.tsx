import React, { useState, useEffect } from 'react';

import { Flex } from 'rebass';

import { Blog } from 'helper/schema';
import BlogTile from './blog-tile';
import Pagination from 'components/pagination';

type Props = {
    blogs: Blog[];
};

/**
 * productPerPage: num of products to be rendered per page.
 * currIndex: current view page num.
 */
type PaginationState = {
    blogPerPage: number;
    currIndex: number;
};

const initialPagination = {
    blogPerPage: 5,
    currIndex: 0,
};

const AllBlogs: React.FC<Props> = ({ blogs }) => {
    // store: all products to be displayed (not filtered by pagination.)
    const [store, setStore] = useState<Blog[]>([]);
    // add product per page
    const [pagination, setPagination] = useState<PaginationState>(
        initialPagination
    );
    const [numOfPages, setNumOfPages] = useState(0);

    // to display products (filtered by pagination)
    const [display, setDisplay] = useState<Blog[]>([]);

    useEffect(() => {
        const duplicated = [...blogs, ...blogs, ...blogs, ...blogs];

        const debug = true;

        setStore(debug ? duplicated : blogs);
    }, []);

    useEffect(() => {
        paginate({ ...pagination }, store);
        calculateNumOfPages({ ...pagination }, store);
    }, [store, pagination]);

    const paginate = (
        { blogPerPage, currIndex }: PaginationState,
        inStore: Blog[]
    ) => {
        const maxIndex = inStore.length;

        const firstIndex = currIndex * blogPerPage;
        const lastIndex = firstIndex + blogPerPage;

        // check if first and/or last index is larger than max and adjust accordingly
        const temp = inStore.slice(
            firstIndex >= maxIndex ? maxIndex : firstIndex,
            lastIndex >= maxIndex ? maxIndex : lastIndex
        );

        setDisplay(temp);
    };

    const calculateNumOfPages = (
        { blogPerPage }: PaginationState,
        inStore: Blog[]
    ) => {
        const initialNumOfPages = inStore.length / blogPerPage;
        setNumOfPages(
            inStore.length % blogPerPage === 0 // check for any rest from the division.
                ? initialNumOfPages
                : Math.floor(initialNumOfPages + 1) // add an extra page if any rest is there.
        );
    };

    const handleClickPage = (pageNum: number) => {
        setPagination(prev => ({
            ...prev,
            currIndex: pageNum,
        }));
    };

    const goToFirst = () => {
        setPagination(prev => ({
            ...prev,
            currIndex: 0,
        }));
    };

    const goToLast = () => {
        setPagination(prev => ({
            ...prev,
            currIndex: numOfPages - 1,
        }));
    };

    return (
        <Flex flexDirection="column" alignItems="center">
            <Flex
                flexWrap="wrap"
                justifyContent="space-between"
                className="content"
                px={[5, 5, 8]}
                width="100%"
                maxWidth="1920px"
                m="0 auto"
            >
                {display.map((blog, i) => (
                    <BlogTile
                        key={`${blog.slug}-${i}`}
                        blog={blog}
                        first={i === 0}
                    />
                ))}
            </Flex>
            <Pagination
                numOfPages={numOfPages}
                current={pagination.currIndex}
                handleClickPage={handleClickPage}
                goToFirst={goToFirst}
                goToLast={goToLast}
            />
        </Flex>
    );
};

export { AllBlogs };
