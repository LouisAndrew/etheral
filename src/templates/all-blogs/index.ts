
import { AllBlogs } from './all-blogs';

export default AllBlogs;