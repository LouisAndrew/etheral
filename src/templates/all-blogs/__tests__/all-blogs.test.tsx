import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

// import { render, cleanup } from '@testing-library/react'
import { cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import { AllBlogs } from '../all-blogs';
import { mockBlog } from 'helper/const';

describe('AllBlogs', () => {
    const Element = <AllBlogs blogs={[mockBlog]} />;

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(Element, div);
    });

    /* it('renders correctly', () => {
		const { getByTestId } = render()
	}) */

    it('matches snapshot', () => {
        const run = true;

        if (run) {
            const tree = renderer.create(Element).toJSON();
            expect(tree).toMatchSnapshot();
        }
    });
});
