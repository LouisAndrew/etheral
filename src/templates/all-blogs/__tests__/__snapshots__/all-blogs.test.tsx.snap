// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllBlogs matches snapshot 1`] = `
<div
  className="css-8i9v14"
>
  <div
    className="content css-2djglc"
  />
  <nav
    className="css-zkfaav"
  >
    <div
      className="css-cfyrb9"
      id="go-to-first"
      onClick={[Function]}
      role="button"
    >
      <svg
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path d=\\"M11 17.035a.5.5 0 0 1-.788.409l-7.133-5.036a.5.5 0 0 1 0-.816l7.133-5.036a.5.5 0 0 1 .788.409v10.07zm1.079-4.627a.5.5 0 0 1 0-.816l7.133-5.036a.5.5 0 0 1 .788.409v10.07a.5.5 0 0 1-.788.409l-7.133-5.036z\\" fill=\\"currentColor\\"/>",
          }
        }
        focusable={false}
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "transform": "rotate(360deg)",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
    <div
      className="css-cfyrb9"
      id="go-to-last"
      onClick={[Function]}
      role="button"
    >
      <svg
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path d=\\"M4.788 17.444A.5.5 0 0 1 4 17.035V6.965a.5.5 0 0 1 .788-.409l7.133 5.036a.5.5 0 0 1 0 .816l-7.133 5.036zM13 6.965a.5.5 0 0 1 .788-.409l7.133 5.036a.5.5 0 0 1 0 .816l-7.133 5.036a.5.5 0 0 1-.788-.409V6.965z\\" fill=\\"currentColor\\"/>",
          }
        }
        focusable={false}
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "transform": "rotate(360deg)",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
  </nav>
</div>
`;
